simplex/comm.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package simplex
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"slices"
+
+	"github.com/ava-labs/simplex"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/message"
+	"github.com/ava-labs/avalanchego/proto/pb/p2p"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/networking/sender"
+	"github.com/ava-labs/avalanchego/subnets"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	_               simplex.Communication = (*Comm)(nil)
+	errNodeNotFound                       = errors.New("node not found in the validator list")
+)
+
+type Comm struct {
+	logger   simplex.Logger
+	subnetID ids.ID
+	chainID  ids.ID
+	// nodeID is this nodes ID
+	nodeID simplex.NodeID
+	// nodes are the IDs of all the nodes in the subnet
+	nodes []simplex.NodeID
+	// sender is used to send messages to other nodes
+	sender     sender.ExternalSender
+	msgBuilder message.OutboundMsgBuilder
+}
+
+func NewComm(config *Config) (*Comm, error) {
+	nodes := make([]simplex.NodeID, 0, len(config.Validators))
+
+	// grab all the nodes that are validators for the subnet
+	for _, vd := range config.Validators {
+		nodes = append(nodes, vd.NodeID[:])
+	}
+
+	if _, ok := config.Validators[config.Ctx.NodeID]; !ok {
+		config.Log.Warn("Node is not a validator for the subnet",
+			zap.String("nodeID", config.Ctx.NodeID.String()),
+			zap.String("chainID", config.Ctx.ChainID.String()),
+			zap.String("subnetID", config.Ctx.SubnetID.String()),
+		)
+		return nil, fmt.Errorf("%w could not find our node: %s", errNodeNotFound, config.Ctx.NodeID)
+	}
+
+	sortNodes(nodes)
+
+	c := &Comm{
+		subnetID:   config.Ctx.SubnetID,
+		nodes:      nodes,
+		nodeID:     config.Ctx.NodeID[:],
+		logger:     config.Log,
+		sender:     config.Sender,
+		msgBuilder: config.OutboundMsgBuilder,
+		chainID:    config.Ctx.ChainID,
+	}
+
+	return c, nil
+}
+
+// sortNodes sorts the nodes in place by their byte representations.
+func sortNodes(nodes []simplex.NodeID) {
+	slices.SortFunc(nodes, func(i, j simplex.NodeID) int {
+		return bytes.Compare(i, j)
+	})
+}
+
+func (c *Comm) ListNodes() []simplex.NodeID {
```

Comment:
I also want to confirm that this is expected to always return all validators, not just the validators we are connected to right? I mainly ask because it seems weird to use the `Communication` interface as a place to expose what validators there are in the network, rather than just who we are currently connected to.